fastcrawler/core/spider.py

```diff
@@ -133,45 +167,65 @@ def add_url_to_pending(self, urls: set[str]) -> None:
                 self.pending_urls.add(url)
         return None
 
-    def pass_url_to_current_spider(self, parsing: HTMLParser) -> None:
+    def pass_url_to_current_spider(self, urls: set[str]) -> None:
         """Pass the url to current spider to crawl them again"""
-        if hasattr(parsing, "resolver") and parsing.resolver:
-            self.add_url_to_pending(parsing.resolver)
+        self.add_url_to_pending(urls)
         return None
 
-    def pass_url_to_next_spider(self, parsing: HTMLParser) -> None:
+    def pass_url_to_next_spider(self, urls) -> None:
         """Pass url to the next spider so that next spider can crawl them"""
-        if hasattr(parsing, "next_resolver") and parsing.next_resolver:
-            for url in parsing.next_resolver:
-                if url not in self.instances[-1].start_url:
-                    self.instances[-1].start_url.add(url)
+        for url in urls:
+            assert type(self.instances[-1].start_url) is set
+            if url not in self.instances[-1].start_url:
+                self.instances[-1].start_url.add(str(url))
         return None
 
-    def parse(self, data: str) -> BaseModel:
+    def parse(self, data: str) -> BaseModel | None:
         """Parse the data from the response w.t.r data model"""
-        parsing: HTMLParser = self.parser(data)
-        result = parsing.parse(self.data_model)
-        self.pass_url_to_current_spider(parsing)
-        self.pass_url_to_next_spider(parsing)
-        return result
+        if self.parser and self._data_model:
+            parsing: HTMLParser = self.parser(data)  # type: ignore
+            # TODO: investigate why HTMLParser doesn't respect proto
+            result = parsing.parse(self._data_model)
+
+            # TODO: handle logic for passing url to next/current spider
+            if hasattr(parsing, "resolver") and parsing.resolver:
+                self.pass_url_to_current_spider(parsing.resolver)  # type: ignore
+                # TODO: use string, not URL
+
+            if hasattr(parsing, "next_resolver") and parsing.next_resolver:
+                self.pass_url_to_next_spider(parsing.next_resolver)
+            return result
 
-    async def save(self, all_data: BaseModel) -> None:
+        return None
+
+    async def save(self, all_data: list[Any | BaseModel | None]) -> None:
         """
         Save the data to somewhere
         Must be implemented in order to save the data
         """
         return None
 
-    def parse_response(self, response: Response) -> BaseModel | None:
+    async def save_cycle(self, all_data: list[RequestCycle]) -> None:
+        """
+        Save the flow of request (request cycle)
+        Must be overwritten if you wish to save more than just the parsed data
+        """
+        await self.save([data.parsed_data for data in all_data if data and data.parsed_data])
+        return None
```

Comment:
dummy return

Reply:
what do you want me to return then?
if you're saying return None should be avoided, then I disagree. There's many good reason you've to explicitly return None. 

Read this article,
then read the comments :^

https://betterprogramming.pub/please-stop-returning-none-from-python-functions-ff98e492b4b0

Reply:
did you actually read the article? it's saying you should avoid returning None
it doesn't make any sense you are calling another method and you are returning None
for absolutely no reason whatsoever.

Reply:
Ofc I read. That’s why I told you to look up in comments, they’ve mentioned why you should return None explicitly